![](https://api.brandy.run/core/core-logo-wide)

# Pandas and Regex practical example

En ese ejercício haremos una simulación. 

Tienes una reunión con tu jefe esa tarde, pero al verificar el dataset por la mañana, hay un problema, ¡nada está en su sítio!

In [1]:
# loading dataframe
import pandas as pd
df = pd.read_csv('db.csv')
df

Empty DataFrame
Columns: [name phone email date contract_value creditcard country postalcode address Cameron Sanders +049 106-232065 In@consequatdolorvitae.co.uk Nov 18,  2018 €1968, 40 5109-0814-0029-5373 Spain 17716 6396 Orci St. Paki Andrews +92 757-624220 sed.tortor.Integer@amet.org Nov 5,  2019 €1436, 12 5193 1959 4162 5216 Spain 85485 669-6354 Orci Ave Alice Henderson +058 588-402901 Phasellus@vitaesodalesnisi.co.uk Mar 9,  2020 €1454, 20 5582.3641.9791.9751 Austria 4964 1696 Ac St. Jamalia Johnston +080 954-388350 turpis.nec@elitafeugiat.net Sep 30,  2019 €1979, 49 5398 2093 5064 8545 Netherlands 7683 XL Ap #958-9660 Dis Rd. Erica Meloni +044 143-705707 pede@risusQuisquelibero.net Sep 24,  2019 €1532, 00 5186 5977 9103 6189 Italy 31046 P.O. Box 271,  1072 Cursus Rd. Hunter Cardenas +042 741-154469 dolor@quis.edu Jul 31,  2020 €1950, 98 5406 4489 0949 9550 United Kingdom C57 2AU 473-2340 Nec Road Anders Bodin +21 119-216844 elit.Aliquam@quistristique.edu Jul 28,  2019 €1763, 38 5511-2605-5014-4845 Sweden 51381 Ap #633-7988 Viverra. St. Karen Chaney +28 655-231748 neque.Morbi.quis@ultriciesadipiscingenim.edu Dec 23,  2018 €1110, 88 5210-8750-6225-4958 Poland 98-144 P.O. Box 733,  6841 Sollicitudin Avenue Ciaran Coleman +009 622-410394 felis.adipiscing.fringilla@facilisis.org Sep 8,  2020 €1199, 90 5508-1717-8235-1477 Austria 5467 P.O. Box 919,  7186 Lectus St. Lily Riviere +048 649-684450 Aenean.egestas.hendrerit@felis.org Aug 15,  2020 €1647, 42 5328-8248-4037-2137 France 67753 Ap #449-9891 Eu,  Street Xaviera Bowen +31 245-397063 sit.amet@maurissit.ca Dec 13,  2019 €1973, 10 5219-8535-9406-4212 Poland 93-445 P.O. Box 907,  4551 Fusce Avenue Samuel Olivieri +033 579-762372 aliquam.arcu@est.co.uk Mar 25,  2019 €1791, 29 5207.5369.2539.2998 Italy 52829 4444 Sed Road Zoe Leroy +95 928-808665 massa@erat.edu Mar 7,  2020 €1640, 89 5210 4758 4547 0615 France 37131 2939 Proin Av. Tatum Middleton +57 791-170967 molestie.arcu.Sed@sitametante.ca Sep 29,  2020 €1415, 83 5410 9173 9811 1391 Belgium 9177 107-2524 Ultricies Ave Sara Bodin +047 305-584900 dictum.Phasellus@lorem.edu Nov 28,  2018 €1013, 09 5253-9920-3166-5145 Sweden 60341 Ap #632-9384 Mauris Avenue Zorita May +17 137-396561 lectus@velitPellentesqueultricies.edu Mar 7,  2020 €1644, 52 5366 9991 8319 2251 Spain 43428 1765 Sed,  Rd. Fitzgerald Davenport +61 604-618037 non@risusquis.com Jul 1,  2020 €1210, 59 5342-8921-5819-0872 Belgium 8062 Ap #711-4012 Eleifend,  St. Nicoletta Rizzo +000 938-458939 lectus.justo@dignissimtemporarcu.co.uk Mar 10,  2020 €1586, 93 5482.7407.5774.5465 Italy 28122 Ap #322-5087 Vitae Road Lana Hahn +009 569-284963 cursus.Integer@accumsan.com Mar 2,  2020 €1140, 46 5286 4021 6930 6158 Poland 43-576 P.O. Box 701,  5874 Auctor,  Rd. Adria Bean +81 358-950542 lobortis.Class@Seddictum.org May 8,  2019 €1777, 45 5285 4548 0100 2555 Austria 2511 Ap #106-488 Ut,  Avenue Sybill Ford +098 286-136669 nec@luctus.co.uk Feb 19,  2019 €1311, 02 5593-7776-8533-9587 Belgium 9321 7398 Nunc Ave Andrew Huff +092 164-186530 magnis@vel.org Sep 7,  2020 €1001, 89 5522 6026 6399 6529 Poland 10-646 5226 Ligula St. Patricia Payne +029 212-403496 condimentum@sitametlorem.co.uk Apr 16,  2020 €1619, 58 5251-2773-9482-8494 Turkey 37999 719-8148 Augue Av. Alexandre Picard +65 502-598783 orci.lacus@facilisisvitae.org Jul 20,  2020 €1921, 60 5322-6769-9179-0814 France 88641 9081 Nisi. Rd. Patience Craig +098 853-074347 lacus@Proinvelit.com Apr 23,  2020 €1231, 89 5441.3698.1914.8136 Spain 97197 7707 Tellus. Av. Isabelle Sloan +072 882-869148 elit@Maurisvelturpis.org Apr 9,  2020 €1973, 53 5351 2640 8962 0930 Germany 28275 362-8631 Ridiculus St. Catarina Persson +69 734-790277 non@Sed.co.uk Mar 2,  2020 €1646, 26 5520 3159 9755 0084 Sweden 71962 964-444 Metus. Road Emerson Wood +085 273-686992 at.sem@nequesedsem.com Mar 30,  2019 €1342, 47 5318 7884 1873 8189 Turkey 38182 Ap #447-2191 Vulputate Av. Bo Foster +063 802-495571 auctor.Mauris.vel@nisi.net May 5,  2020 €1019, 28 5594-

El $becário$ ha cometido un error y la base de datos está corrompida. No hay cópia de seguridad...
Por suerte sabes `regex`! Vamos utilizarlos para recuperar la máxima cantidad posible de datos del fichero corrompido. 

Usa `re.findall` sobre el string abajo y escribe patrones para intentar recuperar lo máximo de datos posibles. Las columnas originales del dataset son: 
- name 
- phone 
- email 
- date 
- contract_value 
- creditcard 
- country 
- postalcode 
- address

In [2]:
import re

In [3]:
with open('db.csv') as file:
    csv = file.read()
print(csv[:600])

name phone email date contract_value creditcard country postalcode address Cameron Sanders +049 106-232065 In@consequatdolorvitae.co.uk Nov 18, 2018 €1968,40 5109-0814-0029-5373 Spain 17716 6396 Orci St. Paki Andrews +92 757-624220 sed.tortor.Integer@amet.org Nov 5, 2019 €1436,12 5193 1959 4162 5216 Spain 85485 669-6354 Orci Ave Alice Henderson +058 588-402901 Phasellus@vitaesodalesnisi.co.uk Mar 9, 2020 €1454,20 5582.3641.9791.9751 Austria 4964 1696 Ac St. Jamalia Johnston +080 954-388350 turpis.nec@elitafeugiat.net Sep 30, 2019 €1979,49 5398 2093 5064 8545 Netherlands 7683 XL Ap #958-9660 Di


name phone email date contract_value creditcard country postalcode address 
Cameron Sanders +049 106-232065 In@consequatdolorvitae.co.uk Nov 18, 2018 €1968,40 5109-0814-0029-5373 Spain 17716 6396 Orci St. 
Paki Andrews +92 757-624220 sed.tortor.Integer@amet.org Nov 5, 2019 €1436,12 5193 1959 4162 5216 Spain 85485 669-6354 Orci Ave 
Alice Henderson +058 588-402901 Phasellus@vitaesodalesnisi.co.uk Mar 9, 2020 €1454,20 5582.3641.9791.9751 Austria 4964 1696 Ac St. 
Jamalia Johnston +080 954-388350 turpis.nec@elitafeugiat.net Sep 30, 2019 €1979,49 5398 2093 5064 8545 Netherlands 7683 XL Ap #958-9660 
Di....

In [55]:
#name
pattern = "[A-Z][a-z]*\s[A-Z][a-z]*\s\+"
names = re.findall(pattern, csv)
names = [n[:-2] for n in names]
len(names)

200

In [14]:
#phone
pattern = "\+\d{2,3}\s\d{3}\-\d{6}"
#pattern = "\+[^A-z]*"
telephones = re.findall(pattern, csv)
# telephones = [t[:-1] for t in telephones]
len(telephones)

200

In [17]:
#email
pattern = "\S+@\S+"
mail = re.findall(pattern,csv)
len(mail)

200

In [20]:
#date
pattern = "[A-Za-z]{3}\s\d{1,2},\s\d{4}"
dates = re.findall(pattern, csv)
len(dates)

200

In [24]:
#contract_value
pattern = "€\d+,\d{2}"
values = re.findall(pattern,csv)
len(values)

200

In [29]:
#creditcard
pattern = "\d{4}\D\d{4}\D\d{4}\D\d{4}"
cc = re.findall(pattern,csv)
len(cc)

200

In [50]:
#country
pattern = "\d{4}\D\d{4}\D\d{4}\D\d{4}\s[A-z\s]+"
country = re.findall(pattern,csv)
country = [re.sub("\d{4}\D\d{4}\D\d{4}\D\d{4}\s", "", c) for c in country]
country = [re.sub("\s[A-z]*$", "", c) for c in country]
len(country)

200

In [64]:
name_pattern = "|".join(names)
country_pattern = "|".join(country)
res = re.split(name_pattern,csv)[1:]
address = [re.split(country_pattern,r)[-1][1:] for r in res]
len(address)

200

In [67]:
address[:5]

['17716 6396 Orci St. ',
 '85485 669-6354 Orci Ave ',
 '4964 1696 Ac St. ',
 '7683 XL Ap #958-9660 Dis Rd. ',
 '31046 P.O. Box 271, 1072 Cursus Rd. ']

In [66]:
import pandas as pd

In [71]:
df = pd.DataFrame({
    "name":names,
    "phone":telephones,
    "email":mail,
    "date":dates,
    "contract_value":values,
    "creditcard":cc,
    "country":country,
    "address":address
})

In [73]:
df.head()

name             phone                             email  \
0   Cameron Sanders  +049 106-232065       In@consequatdolorvitae.co.uk   
1      Paki Andrews   +92 757-624220        sed.tortor.Integer@amet.org   
2   Alice Henderson  +058 588-402901   Phasellus@vitaesodalesnisi.co.uk   
3  Jamalia Johnston  +080 954-388350        turpis.nec@elitafeugiat.net   
4      Erica Meloni  +044 143-705707        pede@risusQuisquelibero.net   

           date contract_value           creditcard      country  \
0  Nov 18, 2018       €1968,40  5109-0814-0029-5373        Spain   
1   Nov 5, 2019       €1436,12  5193 1959 4162 5216        Spain   
2   Mar 9, 2020       €1454,20  5582.3641.9791.9751      Austria   
3  Sep 30, 2019       €1979,49  5398 2093 5064 8545  Netherlands   
4  Sep 24, 2019       €1532,00  5186 5977 9103 6189        Italy   

                                address  
0                  17716 6396 Orci St.   
1              85485 669-6354 Orci Ave   
2                     4964 1696 Ac St.   
3         7683 XL Ap #958-9660 Dis Rd.   
4  31046 P.O. Box 271, 1072 Cursus Rd.

In [74]:
df.to_csv("recovered_data.csv")